5단계 - 로또(수동) 과제 PR 요청드립니다!

taewonh opened the pull request:
로또(수동) 과제 진행이 완료되어 PR 요청 드립니다!
우선 4단계 과제에서 피드백 해주신 항목들에 대해 수정을 먼저 진행하고 5단계 과제를 진행했습니다.

이번 5단계 과제에서 새롭게 추가된 기능이나 변경된 기능은 아래와 같습니다.
- 로또 수동 구매 기능 구현
  - InputView, OutputView 에서 제공하는 기능을 static 하게 사용하도록 변경
    - LottoController 인스턴스 생성시 InputView, OutputView 인스턴스를 주입받는 부분 제거
  - LottoStore 수동 로또 발권 기능 추가
    - 수동 발권할 로또 개수를 입력받아 해당 개수만큼 사용자로부터 로또 번호 입력하는 메소드 제공
    - 위 메소드를 이용하여 발권받은 수동 로또 목록과 총 금액을 입력하여 Lottos 를 제공 받음 (LottoStore.sell 메소드 사용)
    - validate 기능 확장
      - 기존 구매 가능한 개수만 체크하는 부분에 추가로 수동으로 구매하려는 개수가 총 구매 가능 개수를 초과했을 때 예외 추가
- 로또 번호 래퍼 클래스 추가 및 적용
  - 클래스명 : Number
  - 로또 번호 목록 (UniqueNumbers) 에서 List<Number> 의 중복 체크, 정렬 시 사용하기 위해 Number 클래스 내부 메소드(hashCode, compareTo, equals) 재정의 처리)
- 숫자 테스트 시 샘플 데이터를 제공하는 AbstractTest 추가

이번 과제에서 래퍼 클래스를 적용 처리하는 부분으로 인해 변경되는 점이 많아 코드 리딩이 더 까다로울 것으로 예상됩니다..
꾸준한 피드백 항상 감사드리며 이번 과제까지 계속 좋은 피드백 주셔서 너무 너무 감사드립니다!
원철님 덕분에 실력적으로나 코드적으로나 많이 개선되어 감을 느꼈습니다!
마지막 피드백도 잘 부탁드리겠습니다!

taewonh committed:
style(LottoTest): 불필요 인덴트 수정

taewonh committed:
refactor(OutputView): 수익률 1.0 미만인 경우에만 추가 문구 출력 처리

taewonh committed:
test(RankTest): ParameterizedTest 적용

taewonh committed:
docs(README.md): step5 핵심 기능 및 비 기능 목록 작성

taewonh committed:
feat: 로또 수동 구매 기능 구현

- 외부에서 주입받지 않고도 사용하기 위해 InputView, OutputView 기능을 static하게 변경
- LottoStore의 수동 로또 발권 기능 추가
  - 외부에서 manualCount만 입력하면 해당 횟수만큼 번호를 입력받고 발권
- LottoStore의 로또 판매 기능 확장
  - sell 메소드 호출시 기존에 입력하던 자동 구매 금액과 더불에 미리 수동 발권한 로또 목록을 전달하면 나머지 금액에 대해서만 자동 구매 처리
- LottoStore의 validate 기능 확장
  - divide와 manualCount 간 비교 처리 추가

taewonh committed:
refactor(Number): 래퍼 클래스 추가 적용

- 로또 번호 래퍼 클래스 추가 및 적용 (class name : Number)
  - stream을 이용하여 중복 체크 및 정렬을 위한 hashCode, compareTo, equals 재정의 처리

- UniqueNumbers.generate() 시 입력 받은 Number 목록 정렬 수행 처리
  - 기존 NumbersGenerator 에서 번호 자동 생성 시 정렬을 수행하고, 수동 입력 번호는 정렬되지 않는 문제로 인해, UniqueNumbers.generate() 시 정렬을 수행하도록 기능 호출부 통합

- 숫자 테스트 시 샘플 데이터를 재 사용하기 위한 AbstractTest 추가

taewonh committed:
style: 구매 개수 출력 문구 변경

- 자동 및 수동 구매 개수에 대한 문구를 출력하도록 수정
- 자동 및 수동 2가지 중 1가지 Count가 0개인 경우 미 출력

heowc commented on src/main/java/step3/domain/LottoStore.java:
```diff
@@ -2,25 +2,49 @@
+
+    public List<Lotto> generateManualLottos(int manualCount) {
+        if (manualCount == 0) {
+            return new ArrayList<>();
```
`Collections.emptyList()` 를 사용하는 것은 어떨까요? 🤔

taewonh replied:
확실히 emptyList()를 사용해서 메모리 낭비를 줄일 수 있을것 같습니다!

heowc commented on src/main/java/step3/domain/LottoStore.java:
```diff
@@ -2,25 +2,49 @@
+        if (manualCount == 0) {
+            return new ArrayList<>();
+        }
+        InputView.printInputManualNumbers();
```
도메인에서 View가 참조되는 상황인 것 같습니다.
이것을 어떻게 개선해볼 수 있을까요? 🤔

heowc commented on src/main/java/step3/domain/Number.java:
```diff
@@ -0,0 +1,39 @@
+
+import java.util.Objects;
+
+public class Number implements Comparable<Number> {
```
원시값을 포장한 것 아주 좋습니다 👍

heowc commented on src/main/java/step3/domain/UniqueNumbers.java:
```diff
@@ -34,8 +36,8 @@ public boolean isEqualSize(UniqueNumbers numbers) {
-    public boolean contains(int number) {
-        return numbers.contains(number);
+    public boolean contains(Number lottoNumber) {
+        return numbers.stream().anyMatch(number -> number.equals(lottoNumber));
```
`contains`로는 동작하는 않는 걸까요? 🤔

taewonh replied:
앗.. 이 부분은 리팩토링 하는 중간에 반영해놨던걸 미반영 해놔서 누락되었습니다 .. 피드백 감사합니다!

heowc commented on src/main/java/step3/utils/NumbersGenerator.java:
```diff
@@ -13,16 +15,20 @@ public class NumbersGenerator {
     }
+
+    public static UniqueNumbers userInput() {
+        return UniqueNumbers.generate(InputView.inputManualNumbers());
```
유틸 또한 특정 View를 참조하고 있는 것 같습니다.
어떻게 개선해볼 수 있을까요? 🤔

taewonh replied:
해당 피드백과 첫 번째 피드백에서 말씀해주신 View 문제는 Controller에서만 View 영역을 참조하도록 구조를 변경하는 방식으로 처리했습니다!

heowc commented on src/test/java/step3/domain/AbstractTest.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.stream.Collectors;
+import step3.utils.NumbersGenerator;
+
+public class AbstractTest {
```
추상 클래스를 만드셔서 중복코드를 제거해주셨네요 👍 

하지만 테스트가 증가할 수록
- 변수들이 지속적으로 늘어날 것이고
- 변수명을 붙이기 애매하여 뒤에 숫자가 붙는 등 
을 하며 복잡해질 것 같습니다.

또는 테스트로 인해 값이 변경되어 의도치 않게 다른 테스트에 영향을 줄 수 있다고 생각합니다.

어떻게 생각하시나요? 🤔

taewonh replied:
기존 추상 클래스를 제거하고 테스트용 샘플 값을 제공해주는 테스트 전용 유틸 클래스를 생성하는 방식으로 처리했습니다!

기존 테스트 코드에서 공통적으로 사용되는 값의 규칙을 먼저 추출하고
해당 숫자값들을 이용해 테스트 코드에서 원하는 인스턴스를 대신 생성해서 전달해주는 방식입니다.

추가한 기능들은 아래와 같습니다.
1. 입력한 값의 크기에 해당하는 Random UniqueNumbers List를 생성해서 제공하는 기능
2. 입력한 start, end 에 해당하는 UniqueNumbers Instance를 생성해서 제공하는 기능
3. 입력한 start, end 에 대항하는 Number Instance List를 생성해서 제공하는 기능
4. 입력한 숫자값들을 모두 Number로 변환하여 List로 구성해서 제공하는 기능

해당 클래스를 생성하여 피드백에서 말씀해주신 "변수들이 지속적으로 늘어나는 문제", "변수명에 숫자가 붙는 문제" 등에 대해 해결할 수 있을 것으로 보이는데 괜찮은 방법일까요?

heowc replied:
현재로썬 괜찮은 것 같습니다 👍

heowc reviewed (changes requested):
마지막까지 잘 진행해주셨습니다 👍 

노력의 흔적이 여기저기 보이는  것 같습니다.
소소한 코멘트 남겼으니 확인해주세요 👍 

코멘트 보시고 궁금하신 점이나 함께 이야기해보면 좋을 내용이 있다면 깃헙 코멘트 또는 슬랙 DM으로 편하게 말씀 부탁드립니다 🙇

taewonh committed:
refactor: 빈 list 생성 방식 변경

- new ArrayList<>()가 아닌 Collections.emptyList() 를 사용하여 메모리 낭비 최소화

taewonh committed:
refactor(UniqueNumbers): contains 내부 로직 개선

- stream().anyMatch()가 아닌 contains를 사용하여 포함 여부를 확인하도록 변경

taewonh committed:
refactor: 도메인, 유틸 내 View 참조 제거

- 도메인(LottoStore) 및 유틸(NumbersGenerator) 내부에서 참조하는 View 제거
  - LottoStore의 경우 외부에서 manualCount를 입력받는 것이 아닌 Controller에서 이미 생성된 List<UniqueNumbers> 받아 내부적으로 로또 수동 발급 처리
  - NumbersGenerator는 수동 번호 입력 기능을 더이상 제공하지 않고 Controller에서 제공하는 방식으로 변경

taewonh committed:
test(LottoStoreTest): LottoStore.sell 메소드 파라미터 변경 반영

- List<Lotto> 에서 List<UniqueNumbers> 로 파라미터가 변경됨에 따라 테스트 반영

taewonh committed:
test: AbstractTest 제거 및 숫자 관련 기능을 제공하는 테스트용 유틸 클래스 생성

- 추후 확장성을 고려할 때 추상 클래스를 상속받는 방식보다 공통적으로 기능을 제공하는 테스트용 유틸 클래스로 분리
- 현재까지 공통적으로 사용하는 값을 제공하는 기능을 제공하고 외부(테스트 코드)에서 전달하는 값만으로 원하는 것을 생성해서 제공하도록 변경

heowc commented on src/main/java/step3/domain/LottoStore.java:
```diff
@@ -1,26 +1,45 @@
 
     public static final Money pricePerLotto = Money.generate(1000);
 
     public Lottos sell(Money payment) {
```
해당 코드는 이제 필요 없지 않을까요 ? 🤔

heowc commented on src/test/java/step3/utils/TestNumberProvider.java:
```diff
@@ -0,0 +1,39 @@
+
+public class TestNumberProvider {
+
+    public static List<UniqueNumbers> randomUniqueNumbersList(int totalCount) {
```
테스트에서 만들어진 클래스라도 테스트를 작성해주시면 더 좋을 것 같습니다 👍

heowc reviewed (approved):
5단계 마무리까지 잘 진행해주셨어요. 👍
첫번째 미션 진행하느라 고생 많으셨습니다.

피드백을 조금 남겼는데, 반영이 꼭 필요한 부분은 아니라서 여기서 머지하도록 하겠습니다.
남은 미션도 즐겁게 진행하시고, 성장에 도움이 되기를 바라겠습니다.

고생하셨습니다! 🙏
